sbi/samplers/vi/vi_pyro_flows.py

```diff
@@ -0,0 +1,635 @@
+    """This is TransformedDistribution with the capability to return parameters!"""
+
+    __doc__ += torch.distributions.TransformedDistribution.__doc__
+
+    def parameters(self):
+        if hasattr(self.base_dist, "parameters"):
+            yield from self.base_dist.parameters()
+        for t in self.transforms:
+            yield from get_parameters(t)
+
+    def modules(self):
+        if hasattr(self.base_dist, "modules"):
+            yield from self.base_dist.modules()
+        for t in self.transforms:
+            yield from get_modules(t)
+
+
+@docstring_parameter(list(_TRANSFORMS.keys()))
+def build_flow(
+    event_shape: torch.Size,
+    link_flow: transforms.Transform,
+    num_flows: int = 5,
+    transform: str = "affine_autoregressive",
+    permute: bool = True,
+    batch_norm: bool = False,
+    base_dist: Distribution = None,
+    device: str = "cpu",
+    **kwargs,
+) -> TransformedDistribution:
+    """Generates a Transformed Distribution where the base_dist is transformed by
+       num_flows bijective transforms of specified type.
+
+
+
+    Args:
+        event_shape: Shape of the events generated by the distribution.
+        link_flow: Links to a specific support .
+        num_flows: Number of normalizing flows that are concatenated.
+        transform: The type of normalizing flow. Should be one of {0}
+        permute: Permute dimension after each layer. This may helpfull for
+            autoregressive or coupling nets.
+        batch_norm: Perform batch normalization.
+        base_dist: Base distribution. If 'None' then a standard Gaussian is used.
+        device: Device on which we build everythink.
+        kwargs: Hyperparameters are added here.
+    Returns:
+        TransformedDistribution
+
+    """
+    # Some transforms increase dimension by decreasing the degrees of freedom e.g.
+    # SoftMax.
+    additional_dim = len(
+        link_flow(torch.zeros(event_shape, device=device))
+    ) - torch.tensor(event_shape, device=device)
+    event_shape = torch.Size(torch.tensor(event_shape, device=device) - additional_dim)
+    # Base distribution is standard normal if not specified
+    if base_dist is None:
+        base_dist = Independent(
+            Normal(
+                torch.zeros(event_shape, device=device),
+                torch.ones(event_shape, device=device),
+            ),
+            1,
+        )
+    # Generate normalizing flow
+    if isinstance(event_shape, int):
+        dim = event_shape
+    elif isinstance(event_shape, Iterable):
+        dim = event_shape[-1]
+    else:
+        raise ValueError("The eventshape must either be an Integer or a Iterable.")
+
+    flows = []
+    for i in range(num_flows):
+        flows.append(
+            get_transform(transform, dim, device=device, **kwargs).with_cache()
+        )
+        if permute and i < num_flows - 1:
+            permutation = torch.randperm(dim, device=device)
```

Comment:
If I understand this correctly, none of the flows has the `LU` decomposition (originally used in the nsf paper). Maybe add this to the docstring of `spline_coupling` (just to highlight that they are not exactly nsf)

Reply:
I think the implementation is based on Invertible Generative
    Modeling using Linear Rational Splines (https://arxiv.org/pdf/2001.05168.pdf).

Reply:
Fine, but still add to it to the docstring please (also fine to reference this paper there)